src/ign.cc

```diff
@@ -523,10 +540,156 @@ extern "C" IGNITION_FUEL_TOOLS_VISIBLE int downloadUrl(const char *_url,
+
+    if (downloadWorlds)
+    {
+      // Get list of world identifiers in collection
+      auto worldIter = client.Worlds(collection);
+      for (; worldIter; ++worldIter)
+      {
+        worldIds.push_back(worldIter);
+      }
+      ignmsg << "Found " << worldIds.size() << " worlds in collection ["
+        << collection.Name() << "]" << std::endl;
+    }
+
+    const std::size_t totalItemCount = modelIds.size() + worldIds.size();
+    if (totalItemCount == 0)
+    {
+      std::cout << "There are no items in collection [" << collection.Name()
+        << "]" << std::endl;
+      return false;
+    }
+
+    size_t jobs = std::max(1, _jobs);
+
+    ignmsg << "Using " << jobs << " jobs to download collection of "
+           << totalItemCount << " items" << std::endl;
+
+    std::deque<std::future<ignition::fuel_tools::Result>> tasks;
+
+    // Check for finished tasks by checking if the status of their futures is
+    // "ready". If a task is finished, check if it succeeded and print out an
+    // error message if it failed. When a task is finished, it gets erased from
+    // the tasks list to make room for other tasks to be added.
+    size_t itemCount = 0;
+    auto checkForFinishedTasks = [&itemCount, &totalItemCount, &tasks] {
+      auto finishedIt =
+          std::partition(tasks.begin(), tasks.end(), [](const auto &_task)
+              {
+                return std::future_status::ready !=
+                _task.wait_for(std::chrono::milliseconds(100));
+              });
+
+      if (finishedIt != tasks.end())
+      {
+        for (auto taskIt = finishedIt; taskIt != tasks.end(); ++taskIt)
+        {
+          ignition::fuel_tools::Result result = taskIt->get();
+          if (result)
+          {
+            ++itemCount;
+          }
+          else
+          {
+            ignerr << result.ReadableResult() << std::endl;
+          }
+        }
+
+        tasks.erase(finishedIt, tasks.end());
+        ignmsg << "Downloaded: " << itemCount << " / " << totalItemCount
+               << std::endl;
+      }
+    };
+
+    // Here we use std::async to download items in parallel. The download task
+    // is started asynchronously and gets added to the task list which is
+    // monitored for completion.
+    if (downloadModels)
+    {
+      for (const auto &modelId : modelIds)
+      {
+        // Check if any of the tasks are done. Don't start a new task until the
+        // number of tasks in the tasks lists is below the number of jobs
+        // specified by the user.
+        while (tasks.size() >= jobs)
+        {
+          checkForFinishedTasks();
+        }
+        std::this_thread::sleep_for(std::chrono::milliseconds(10));
+        auto handle = std::async(std::launch::async, [&modelId, &client]
+            {
+              return client.DownloadModel(modelId);
+            });
+        tasks.push_back(std::move(handle));
+      }
+    }
+
+    if (downloadWorlds)
+    {
+      for (auto &worldId : worldIds)
```

Comment:
```suggestion
      for (const auto &worldId : worldIds)
```

Reply:
We actually need a mutable reference here because `DownloadWorld` updates the id. Left a comment saying so in https://github.com/ignitionrobotics/ign-fuel-tools/pull/98/commits/5a90f5a36adee164ccdb1fcb2a3622f561cb61b5

Reply:
Oh gotcha. Interesting that the behaviour is different from `DownloadModel` :confused: